### PR TITLE
Solver optimizations and cleanup

### PR DESCRIPTION
**Pull request by kburns**
This PR includes a few solver optimizations, including:
* Combining matvecs for the subsystems in the same subproblem
* Combining linear solvers for the subsystems in the same subproblem

and a lot of solver class cleanup and rounding out a few features, including:
* Adding v2-style left eigenvector support the sparse EVP
* Adding `rebuild_matrices` options to the LBVP/EVP to support on-the-fly field NCC modifications
* Allowing you to specify specific subsystems for the LBVP like for the EVP (addressing #218).

It would be nice to add some left eigenvector tests / examples.

### CODE REVIEW

**Comment by kburns** [comment]
Added big clean up of the problem classes, including adding back in all equation condition checking and sensible error messages when problems are detected.

File: dedalus/core/problems.py [hunk 1]
```diff
@@ -45,281 +33,134 @@
...
 
     def __init__(self, variables, namespace=None):
         self.variables = variables
         self.LHS_variables = variables
         self.dist = unify_attributes(variables, 'dist')
         self.equations = self.eqs = []
         # Build namespace via chainmap to react to upstream changes
-        # Priority: local dict, specified namespace, parseables
+        # Priority: local namespace, external namespace, parseables
+        self.local_namespace = {}
```
Comment (lecoanet) [inline dedalus/core/problems.py:46]: Why do we make a local_namespace? Is this so someone could modify this after the problem instantiation? I don't see this used anywhere right now.

**Comment by lecoanet** [comment]
Should we rewrite the RBC EVP example to use the rebuild_matrices? It is a bit complicated because kx isn't a parameter, it is part of the field. Maybe we could make it a 1D problem and define a custom grad operator? Sounds unnecessarily complicated....

File: dedalus/core/problems.py (hunk 1 above)
Comment (kburns) [reply dedalus/core/problems.py:46]: This is so that you can directly add things to the namespace after instantiating the problem without polluting locals, i.e. `problem.namespace['a'] = A` will make `a` available for equation parsing, but won't create/change `a` in the script namespace.

**Comment by kburns** [comment]
Yeah it would be great to have a better way to handle that eventually.  Maybe once we have domain remapping / metric tensors implemented, you could just modify the horizontal stretching parametrically and rebuild matrices.  But for now this is probably the cleanest thing besides writing the equations scalar/component style like in v2.

**Comment by lecoanet** [comment]
I tested performance on an Nr=256, Lmax=127 problem on Pleiades with 1024 cores (mesh=[32, 32]). There is some run-to-run variability which I think is due to differences in parallel transpose times. The new PR does not affect performance within the run-to-run variability. However, I looked at how much time is actually spent on the matrix solve and the csr_matvec/csr_matvecs. I measured these relative to the amount of time spent on attempt_tasks, which I think should be fairly consistent across simulations because it does not depend on the transpose timings.

The new matrix solves are very slightly faster, by about 2%. That's to say that if before the solves were 3% of the simulation time, now they are 2.94% of the simulation time. However, the new csr_matvecs are more robustly slower, by about 30%. They've gone from ~1.5% of the simulation time to ~1.95% of the simulation time. Overall, the simulations are slower with the changes made in this PR.

File: dedalus/core/solvers.py [hunk 2]
```diff
@@ -513,7 +516,7 @@ def __init__(self, problem, timestepper, enforce_real_cadence=100, warmup_iterat
         # Initialize timestepper
         self.timestepper = timestepper(self)
         # Attributes
-        self.sim_time = self.initial_sim_time = 0  # TODO: allow picking up from current problem sim time?
+        self.sim_time = self.initial_sim_time = problem.time['g'].ravel()[0]
```
Comment (lecoanet) [inline dedalus/core/solvers.py:519]: This doesn't work in parallel.

**Comment by lecoanet** [comment]
More confusion on performance... I tried testing the internally heated convection problem on 4 cores on my laptop. There it seems like the new solves are faster by ~15% and csr_matvecs are faster by a factor of 2. However, running the problem on the PR is consistently slower (by 0.5%) than on master. Digging down, I noticed a big difference in the amount of time spent in "concatenate" call on increment_group and decrement_group. On master, this takes 4.79% of the simulation time, whereas on the PR it takes 12.79% of the simulation time. The weird thing is there doesn't seem to be any difference in concatenate timing on Pleiades. Could there be something about data allocation/ordering in the new gather/scatter that makes the concatenate call on increment_group and decrement_group slower?

File: dedalus/core/solvers.py (hunk 2 above)
Comment (kburns) [reply dedalus/core/solvers.py]: Fixed

**Comment by kburns** [comment]
I'm not sure, maybe it's interacting with load balancing in some strange way? Those methods are directly concatenating all local data, not going through the subsystem/subproblem gather or scatter interfaces.

**Comment by lecoanet** [comment]
I've now also tested performance on my desktop (mac studio with new mac chip). On the internally heated convection problem (on 4 cores), I find the new csr_matvecs are _faster_ than the implementation on master by ~x2, similar to my laptop. The solves are about the same speed. Overall the simulation seems to be ~3% faster on the PR, although there is some run-to-run variability. I also tested the 2D rayleigh benard problem (on 4 cores). There I found the csr_matvecs are _slower_ than the implementation on master by ~x2; the solves are similar. Mostly due to the slow matvecs's, the 2D RBC on the PR were 15% slower than on master.

To summarize the results of these various tests: It seems like the new matrix solves are similar in performance to the old matrix solves. The new matvecs's are sometimes much faster, sometimes much slower than the old ones. If we want to improve performance, I would suggest that we make a configuration option where the user can choose whether to use individual matvec's for each subsystem, or do one big matvecs for the whole subproblem.

**Comment by kburns** [comment]
Seems very non-reproducible. On my mac laptop, the 2D RBC example is 15% faster on the PR than on master, entirely because csr_matvecs is 3x faster than csr_matvec.

File: dedalus/core/problems.py [hunk 3]
```diff
@@ -329,201 +170,165 @@ def _build_matrix_expressions(self, eqn):
...
 
-    @CachedAttribute
-    def namespace_additions(self):
-        """Build namespace for problem parsing."""
-        # Variable perturbations
-        self.perturbations = [Field(bases=var.bases, name='δ'+var.name, dtype=var.dtype) for var in self.variables]
-        return {pert.name: pert for pert in self.perturbations}
+    def _check_equation_conditions(self, eqn):
+        """Check equation conditions."""
+        # No conditions
```
Comment (lecoanet) [inline dedalus/core/problems.py:231]: Do we need to check linearity of LHS?

Comment (lecoanet) [reply dedalus/core/problems.py:231]: Maybe also domain containment?

File: dedalus/core/solvers.py [hunk 4]
```diff
@@ -144,99 +138,112 @@ class EigenvalueSolver(SolverBase):
...
-
+        subproblem : Subproblem object
+            Subproblem for which to solve the EVP.
+        rebuild_matrices : bool, optional
+            Rebuild LHS matrices if coefficients have changed (default: False).
+        left : bool, optional
+            Solve for the left eigenvectors of the system in addition to the
+            right eigenvectors. The left eigenvectors are the right eigenvectors
+            of the conjugate-transposed problem. Follows same definition described
+            in scipy.linalg.eig documentation (default: False).
```
Comment (lecoanet) [inline dedalus/core/solvers.py:180]: Should we mention we're multiplying by -M? I don't think that is part of the scipy.linalg.eig documentation.

File: dedalus/core/problems.py (hunk 3 above)
Comment (kburns) [reply dedalus/core/problems.py:231]: We don't need them since the current NLBVP setup takes `LHS = RHS` and converts it to a straight nonlinear root-finding using both sides: `H(X) = LHS - RHS = 0`, so there's no restrictions on LHS vs RHS.

**Comment by lecoanet** [comment]
On Pleiades, the 2D RBC problem on 4 cores is ~5% faster on master than on the PR. I think most of the difference is csr_matvecs takes 50% more time on the PR. I don't have any evidence that the performance is non-reproducible, but it does seem to be different on different problems and different machines. For many of the cases I've tested, it seems to make things slower, so I don't think we should remove the old implementation from the code.

**Comment by kburns** [comment]
Can you tell from your examples if the speed differences are in the `_sparsetools.csr_matvec(s)` or just `array.csr_matvec(s)`?  If the former, then I think we can just put the fallback option inside of `array.csr_matvecs` so we don't have to pollute all the solver/timestepper code with extra subsystem loops.

**Review by lecoanet — COMMENTED** [review]
I've gone through the PR and it looks good. The error messages around equation entry are now much more expressive!! There is still an opaque error message that can come up: when you accidentally add a scalar to a vector. You get an error saying "ValueError: Objects are not all equal." which may be hard to interpret. I guess it is not specifically an issue with equation entry, as it can happen when you form any expression. But I think it would be useful to make this clearer. Otherwise I think it is good to merge after we resolve the performance issues with the matvec's.

**Comment by lecoanet** [comment]
The speed differences appear to be in _sparsetools.csr_matvec(s). I think you're right that putting the options in array.csr_matvecs might be a good solution.

**Comment by kburns** [comment]
@lecoanet does it look like this config option enough to get the pre-PR performance on your machines?

**Comment by lecoanet** [comment]
I tested the 2D RBC problem on Pleiades with 4 cores. Before the _sparsetools.csr_matvecs was spending ~50% more time than the _sparsetools.csr_matvec calls on the master branch, but the time spent in array's csr_matvec(s) calls were similar. With SPLIT_CSR_MATVECS = True, the same amount of time is spent in _sparetools.csr_matvec as on master, but now significantly more time in spent in array's csr_matvec. Overall, the extra time spent in array means that slightly more time is spent on the matvec's with SPLIT_CSR_MATVECS = True. That means the matvec's are still ~50% slower on the PR than on master, even when using the configuration option. I can check to see if it helps things at all on my desktop when I get back to Chicago in a couple of weeks.

**Comment by kburns** [comment]
Do you mean array.csr_matvecs? With SPLIT_CSR_MATVECS on the PR, array.csr_matvec shouldn't be called at all, it should directly call _sparsetools.csr_matvec from array.csr_matvecs.

**Comment by kburns** [comment]
Also what does the difference in overall runtimes (without profiling) look like now?  The differences between master and with/without splitting on the branch are now lower than the run-to-run variations on both my linux desktop and the flatiron clusters.

**Comment by kburns** [comment]
I think we should just merge this in for now since we have the config option, and there's lots of other nice API things we'd like to continue building on, but happy to work on more performance analysis and any necessary fixes in another PR.

**Comment by jsoishi** [comment]
+1
